source/grid/tria.cc

```diff
@@ -5177,6 +5179,1237 @@ namespace internal
+                for (const auto i : quad->vertex_indices())
+                  vertex_indices[k++] = quad->vertex_index(i);
+
+                for (const auto i : quad->line_indices())
+                  vertex_indices[k++] =
+                    quad->line(i)->child(0)->vertex_index(1);
+
+                vertex_indices[k++] = next_unused_vertex;
+              }
+
+              boost::container::small_vector<
+                typename Triangulation<dim, spacedim>::raw_line_iterator,
+                12>
+                lines(quad->n_lines() * 3);
+
+              {
+                unsigned int k = 0;
+
+                for (unsigned int l = 0; l < quad->n_lines(); ++l)
+                  for (unsigned int c = 0; c < 2; ++c)
+                    {
+                      static constexpr std::array<std::array<unsigned int, 2>,
+                                                  2>
+                        index = {
+                          {{{1, 0}}, // child 0, line_orientation=false and true
+                           {{0,
+                             1}}}}; // child 1, line_orientation=false and true
+
+                      lines[k++] = quad->line(l)->child(
+                        index[c][quad->line_orientation(l)]);
+                    }
+
+                for (unsigned int l = 0; l < new_lines.size(); ++l)
+                  lines[k++] = new_lines[l];
+              }
+
+              boost::container::small_vector<int, 12> line_indices(
+                lines.size());
+              for (unsigned int i = 0; i < line_indices.size(); ++i)
+                line_indices[i] = lines[i]->index();
+
+              static constexpr std::array<std::array<unsigned int, 2>, 12>
+                line_vertices_quad{{{{0, 4}},
+                                    {{4, 2}},
+                                    {{1, 5}},
+                                    {{5, 3}},
+                                    {{0, 6}},
+                                    {{6, 1}},
+                                    {{2, 7}},
+                                    {{7, 3}},
+                                    {{6, 8}},
+                                    {{8, 7}},
+                                    {{4, 8}},
+                                    {{8, 5}}}};
+
+              static constexpr std::array<std::array<unsigned int, 4>, 4>
+                quad_lines_quad{{{{0, 8, 4, 10}},
+                                 {{8, 2, 5, 11}},
+                                 {{1, 9, 10, 6}},
+                                 {{9, 3, 11, 7}}}};
+
+              static constexpr std::
+                array<std::array<std::array<unsigned int, 2>, 4>, 4>
+                  quad_line_vertices_quad{
+                    {{{{{0, 4}}, {{6, 8}}, {{0, 6}}, {{4, 8}}}},
+                     {{{{6, 8}}, {{1, 5}}, {{6, 1}}, {{8, 5}}}},
+                     {{{{4, 2}}, {{8, 7}}, {{4, 8}}, {{2, 7}}}},
+                     {{{{8, 7}}, {{5, 3}}, {{8, 5}}, {{7, 3}}}}}};
```

Comment:
Would you mind adding some ascii art here to show where these numbers come from? It's really hard to understand this without some documentation.

Reply:
These tables are determined based on the way a Triangulation orders the subobjects - e.g., the quads are ordered in a way that is compatible with the way a Triangulation sets up lines. Hence there really is no better way to explain where these come from, so drawing pictures really won't help that much.